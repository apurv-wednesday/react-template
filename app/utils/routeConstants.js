export default {
  repos: {
    route: '/',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  itunes: {
    route: '/itunes',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  trackDetails: {
    route: '/itunes/:trackId',
    exact: true
  }
};
