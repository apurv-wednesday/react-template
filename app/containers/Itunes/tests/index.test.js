/**
 *
 * Tests for Itunes container
 *
 *
 */

import React from 'react';
// import { fireEvent } from '@testing-library/dom';
import { timeout, renderProvider } from '@utils/testUtils';
import { ItunesTest as Itunes } from '../index';
import { fireEvent } from '@testing-library/dom';
import { translate } from '@app/utils';

describe('<Itunes /> container tests', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<Itunes dispatchSongsApi={submitSpy} />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should call dispatchClearSongsData on empty change', async () => {
    const getSongsApiSpy = jest.fn();
    const clearSongsDataSpy = jest.fn();
    const { getByTestId } = renderProvider(
      <Itunes dispatchClearSongsData={clearSongsDataSpy} dispatchSongsApi={getSongsApiSpy} />
    );
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: 'a' }
    });
    await timeout(500);
    expect(getSongsApiSpy).toBeCalled();
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: '' }
    });
    await timeout(500);
    expect(clearSongsDataSpy).toBeCalled();
  });
  it('should call dispatchSongsApi on change and after enter', async () => {
    const songName = 'coldplay';
    const { getByTestId } = renderProvider(<Itunes dispatchSongsApi={submitSpy} />);
    const searchBar = getByTestId('search-bar');
    fireEvent.change(searchBar, {
      target: { value: songName }
    });
    await timeout(500);
    expect(submitSpy).toBeCalledWith(songName);

    fireEvent.keyDown(searchBar, {
      key: 'Enter',
      code: 13,
      charCode: 13
    });
    expect(submitSpy).toBeCalledWith(songName);
  });
  it('should call dispatchSongsApi on clicking the search icon', async () => {
    const songName = 'coldplay';
    const { getByTestId } = renderProvider(<Itunes dispatchSongsApi={submitSpy} songName={songName} />);
    fireEvent.click(getByTestId('search-icon'));

    await timeout(500);
    expect(submitSpy).toBeCalledWith(songName);
  });

  it('should  dispatchSongsApi on update on mount if songName is already persisted', async () => {
    const songName = 'react-template';
    renderProvider(<Itunes songName={songName} songsData={null} dispatchSongsApi={submitSpy} />);

    await timeout(500);
    expect(submitSpy).toBeCalledWith(songName);
  });

  // it('should validate mapDispatchToProps actions', async () => {
  //   const dispatchReposSearchSpy = jest.fn();
  //   const songName = 'coldplay';
  //   const actions = {
  //     dispatchGithubRepos: { songName, type: homeContainerTypes.REQUEST_GET_GITHUB_REPOS },
  //     dispatchClearGithubRepos: { type: homeContainerTypes.CLEAR_GITHUB_REPOS }
  //   };
  //   const props = mapDispatchToProps(dispatchReposSearchSpy);
  //   props.dispatchGithubRepos(songName);
  //   expect(dispatchReposSearchSpy).toHaveBeenCalledWith(actions.dispatchGithubRepos);

  //   await timeout(500);
  //   props.dispatchClearGithubRepos();
  //   expect(dispatchReposSearchSpy).toHaveBeenCalledWith(actions.dispatchClearGithubRepos);
  // });
  it('should render default error message when search goes wrong', () => {
    const defaultError = translate('something_went_wrong');
    const { getByTestId } = renderProvider(<Itunes songsError={defaultError} />);
    expect(getByTestId('error-message')).toBeInTheDocument();
    expect(getByTestId('error-message').textContent).toBe(defaultError);
  });
  it('should render the default message when searchBox is empty and songsError is null', () => {
    const defaultMessage = translate('repo_search_default');
    const { getByTestId } = renderProvider(<Itunes />);
    expect(getByTestId('default-message')).toBeInTheDocument();
    expect(getByTestId('default-message').textContent).toBe(defaultMessage);
  });
  it('should render the data when loading becomes false', () => {
    const songsData = { items: [{ repoOne: 'react-template' }] };
    const { getByTestId } = renderProvider(<Itunes songsData={songsData} dispatchSongsApi={submitSpy} />);
    expect(getByTestId('for')).toBeInTheDocument();
  });
  it('should render Skeleton Comp when "loading" is true', async () => {
    const songName = 'coldplay';
    const { getByTestId, getAllByTestId } = renderProvider(
      <Itunes loading dispatchSongsApi={submitSpy} songName={songName} />
    );
    expect(getAllByTestId('skeleton').length).toBe(3);
  });
});
