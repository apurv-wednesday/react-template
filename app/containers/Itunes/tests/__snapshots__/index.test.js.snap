// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Itunes /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="css-6azwsi e1n6u2mw1"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root e1n6u2mw3 css-19cof65-MuiPaper-root-MuiCard-root"
        maxwidth="500"
      >
        <div
          class="MuiCardHeader-root e1n6u2mw2 css-1e709b4-MuiCardHeader-root"
        >
          <div
            class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
          >
            <span
              class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
            >
              Search iTunes songs
            </span>
          </div>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth MuiDivider-light css-b2j1b7-MuiDivider-root"
        />
        <p
          class="css-darbe1 egtqi0h0"
          data-testid="t"
        >
          Search Songs
        </p>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedEnd e1n6u2mw0 css-16qisbh-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="search-bar"
            placeholder="wednesday-solutions/react-template"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="search songs"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="search-icon"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root e1n6u2mw3 css-18bnhwu-MuiPaper-root-MuiCard-root"
        color="grey"
      >
        <div
          class="MuiCardHeader-root e1n6u2mw2 css-1e709b4-MuiCardHeader-root"
        >
          <div
            class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
          >
            <span
              class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
            >
              Repository List
            </span>
          </div>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth MuiDivider-light css-b2j1b7-MuiDivider-root"
        />
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="default-message"
        >
          Search for a song by entering song name in the search box
        </p>
      </div>
    </div>
  </div>
</body>
`;
